core/frontend/src/extension/Extension.ts

```diff
@@ -32,47 +37,101 @@ export interface ExtensionManifest {
    * e.g "./lib/main.js"
    */
   readonly main: string;
-  /** List of activation events this Extension supports. */
-  readonly activationEvents: ActivationEvent[];
-}
-
-/** @alpha */
-export interface BuildExtensionManifest extends ExtensionManifest {
   /** Only valid when the Extension is loaded at build-time.
    *
    * Defines the main ES module that will be imported
    */
-  readonly module: string;
+  readonly module?: string;
+  /** List of activation events this Extension supports. */
+  readonly activationEvents: ActivationEvent[];
 }
 
-/** Describes an Extension that has already been downloaded and has a location files can be easily executed.
+/**
+ * A "ready to use" Extension (contains a manifest object and a function to execute).
+ * Will be used as the type for in-memory extensions in the ExtensionAdmin
  * @alpha
-*/
-export interface LocalExtensionProps {
-  readonly manifest: ExtensionManifest;
-  readonly mainFunc?: ResolveFunc;
+ */
+export interface InstalledExtension {
+  /** A function that executes the main entry point of the extension */
+  execute: ResolveFunc;
+  /** The manifest (package.json) of the extension */
+  manifest: ExtensionManifest;
 }
 
-/** The minimum information required to download an Extension
+/**
+ * An Extension Provider defines how to fetch and execute an extension.
+ * An extension can be one of three kinds:
+ *   1. A locally installed extension.
+ *   2. A remote extension served from a user provided host.
+ *   3. A remote extension served from Bentley's Extension Service.
+ * All three must have these required properties and methods.
  * @alpha
-*/
-export interface ExtensionLoaderProps {
-  name: string;
-  version: string;
+ */
+export interface ExtensionProvider {
+  /** A function that returns the extension's manifest file */
+  getManifest: ResolveManifestFunc;
+  /** A function that executes the main entry point of the extension */
+  execute: ResolveFunc;
```

Comment:
Why do we have the `execute` on both the ExtensionProvider and InstalledExtension? It seems like it'd only be necessary on the `InstalledExtension`.

Edit: I see why there's two now since the `execute` on the ExtensionProvider is really the "download" function that we had previously and is just in charge of getting that entry module but not technically executing it...

Reply:
Technically, when execute() on the ExtensionProvider is called, the code is downloaded and executed at the same time. Actually there is no concept of a difference between "download" vs "execute" anymore. I understand the confusion... originally I put the whole Provider inside the InstalledExtension like so `{ provider: ExtensionProvider, manifest: ExtensionManifest }`. But now when ExtensionAdmin wants to execute the extension its more verbose: `extension.provider.execute()` instead of just `extension.execute()`. Thought it was better to get rid of providers at this point since they've already done their job

Reply:
Yea I guess that's where I was getting mixed up, there isn't a difference between the two of them. However, I wonder if we do just keep the `ExtensionProvider` in the `InstalledExtension` to make it explicit. Especially since it's just an internal concept of the `ExtensionAdmin` the added info isn't too bad.

In addition, I think the `ExtensionProvider` would end up not being cleaned up anyway since it would own the execute method that's set as the execute method.